src/submission/submission-to-clinical/submission-to-clinical.ts

```diff
@@ -60,117 +56,102 @@ import * as messenger from '../submission-updates-messenger';
+		// confirm that the state is VALID
+		throw new Errors.StateConflict(
+			'Active submission does not have state PENDING_APPROVAL and cannot be committed.',
+		);
+	} else {
+		// We Did It! We have a valid active submission to commit! Everyone cheers!
+		await performCommitSubmission(activeSubmission);
+	}
 };
 
 const performCommitSubmission = async (
-  activeSubmission: DeepReadonly<ActiveClinicalSubmission>,
+	activeSubmission: DeepReadonly<ActiveClinicalSubmission>,
 ) => {
-  // Get all the current donor documents
-  const donorDTOs = await getDonorDTOsForActiveSubmission(activeSubmission);
-  if (donorDTOs === undefined || _.isEmpty(donorDTOs)) {
-    throw new Errors.StateConflict(
-      'Donors for this submission cannot be found in the clinical database.',
-    );
-  }
-
-  // Update with all relevant records
-  const updatedDonorDTOs = await mergeActiveSubmissionWithDonors(activeSubmission, donorDTOs);
-
-  const verifiedDonorDTOs: Donor[] = [];
-  // check donor if was invalid against current dictionary
-  const currentDictionary = await dictionaryManager.instance().getCurrent();
-  for (const ud of updatedDonorDTOs) {
-    if (ud.schemaMetadata.isValid === false) {
-      L.debug('Donor is invalid, revalidating if valid now');
-      const isValid = dictionaryManager.revalidateAllDonorClinicalEntitiesAgainstSchema(
-        ud,
-        currentDictionary,
-      );
-      if (isValid) {
-        L.info(`donor ${ud._id} is now valid`);
-        ud.schemaMetadata.isValid = true;
-        ud.schemaMetadata.lastValidSchemaVersion = currentDictionary.version;
-      }
-    }
-    // recalculate the donors stats
-    verifiedDonorDTOs.push(recalculateDonorStatsHoldOverridden(ud));
-  }
-
-  try {
-    // write each updated donor to the db
-    await donorDao.updateAll(verifiedDonorDTOs.map(dto => F(dto)));
-
-    // If the save completed without error, we can delete the active registration
-    submissionRepository.deleteByProgramId(activeSubmission.programId);
-
-    sendMessageOnUpdatesFromClinicalSubmission(activeSubmission);
-  } catch (err) {
-    throw new Error(`Failure occured saving clinical data: ${err}`);
-  }
+	// Get all the current donor documents
+	const donorDTOs = await getDonorDTOsForActiveSubmission(activeSubmission);
+	if (donorDTOs === undefined || _.isEmpty(donorDTOs)) {
+		throw new Errors.StateConflict(
+			'Donors for this submission cannot be found in the clinical database.',
+		);
+	}
+
+	// Update donors with submission data
+	//  1. merge submission data into donor documents
+	//  2. validate updated donors vs current schema
+	//  3. recalculate donors core complete stats
+	const mergedDonors = await mergeActiveSubmissionWithDonors(activeSubmission, donorDTOs);
+	const donorsWithSchemaValidation = await validateDonorsWithCurrentDictionary(mergedDonors);
+	const updatedDonors = await updateDonorsCompletionStats(donorsWithSchemaValidation);
```

Comment:
Here we can see the donor validation and core completion updates happening in separate steps to the entire donor array instead of being applied individually to each donor.